parquet_file/src/rebuild.rs

```
@@ -0,0 +1,566 @@
+                catalog.revision_counter() + 1,
+                revision_counter,
+                "revision counter during transaction simulation out-of-sync"
+            );
+
+            if let Some((uuid, entries)) = revisions.get(&revision_counter) {
+                // we have files for this particular transaction
+                let mut transaction = catalog.open_transaction_with_uuid(*uuid).await;
+                for (path, metadata) in entries {
+                    let path: DirsAndFileName = path.clone().into();
+                    transaction
+                        .add_parquet(&path, metadata)
+                        .context(FileRecordFailure)?;
+                }
+                transaction.commit().await.context(CommitFailure)?;
+            } else {
+                // we do not have any files for this transaction (there might have been other actions though or it was
+                // an empty transaction) => create new empty transaction
+                let transaction = catalog.open_transaction().await;
+                transaction.commit().await.context(CommitFailure)?;
+            }
+        }
+    }
+
+    Ok(catalog)
+}
+
+/// Collect all files under the given locations.
+///
+/// Returns a map of revisions to their UUIDs and a vector of file-metadata tuples.
+///
+/// The file listing is recursive.
+async fn collect_revisions(
+    object_store: &ObjectStore,
+    search_location: &Path,
+) -> Result<HashMap<u64, (Uuid, Vec<(Path, ParquetMetaData)>)>> {
+    let mut stream = object_store
+        .list(Some(search_location))
+        .await
+        .context(ReadFailure)?;
+
+    // revision -> (uuid, [file])
+    let mut revisions: HashMap<u64, (Uuid, Vec<(Path, ParquetMetaData)>)> = HashMap::new();
+
+    while let Some(paths) = stream.try_next().await.context(ReadFailure)? {
+        for path in paths.into_iter().filter(is_parquet) {
+            let (iox_md, parquet_md) = read_parquet(object_store, &path).await?;
```

I wonder if it would be better to skip the file if metadata can't be read rather than failing the rebuild. I am thinking of a case where one file got corrupted somehow so its metadata can't be read. It might be valuable to be able to recreate the catalog (automatically) without that one file.

---

The user could also delete the file, since rebuilding the catalog is a manual intervention anyways. However I'm also fine with ignoring affected files as long as this is controlled via a flag (off by default in the upcoming CLI) since I think we should not assume the user intends to ignore errors by default.

---

good points -- I am also fine either way

---

added a flag (+ test)